storage/storage-interface/src/async_proof_fetcher.rs

```diff
@@ -0,0 +1,143 @@
+    // in a single thread.
+    fn finish_and_read_proofs(&self) -> HashMap<HashValue, SparseMerkleProof> {
+        let mut proofs = HashMap::new();
+        for _ in 0..self.num_proofs_to_read.load(Ordering::Relaxed) {
+            let data = self
+                .data_receiver
+                .recv()
+                .expect("Failed to receive proof on the channel.");
+            match data {
+                Proof {
+                    state_key_hash,
+                    proof,
+                } => {
+                    proofs.insert(state_key_hash, proof);
+                }
+            }
+        }
+        self.num_proofs_to_read.store(0, Ordering::Relaxed);
+        proofs
+    }
+
+    // Schedules proof reading work in a background running thread pool.
+    fn read_proof_async(&self, state_key: StateKey, version: Version) {
+        self.num_proofs_to_read.fetch_add(1, Ordering::Relaxed);
+        let reader = self.reader.clone();
+        let data_sender = self.data_sender.clone();
+        IO_POOL.spawn(move || {
+            let proof = reader
+                .get_state_proof_by_version(&state_key, version)
+                .expect("Proof reading should succeed.");
+            data_sender
+                .send(Proof {
+                    state_key_hash: state_key.hash(),
+                    proof,
+                })
+                .expect("Sending proof should succeed.");
+        });
+    }
+}
+
+impl ProofFetcher for AsyncProofFetcher {
+    fn fetch_state_value_and_proof(
+        &self,
+        state_key: &StateKey,
+        version: Version,
+    ) -> Result<(Option<StateValue>, Option<SparseMerkleProof>)> {
+        self.read_proof_async(state_key.clone(), version);
+        let value = self.reader.get_state_value_by_version(state_key, version)?;
+        Ok((value, None))
+    }
+
+    fn get_proof_cache(&self) -> HashMap<HashValue, SparseMerkleProof> {
+        self.finish_and_read_proofs()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::mock::MockDbReaderWriter;
+    use assert_unordered::assert_eq_unordered;
+
+    #[test]
+    fn test_fetch() {
+        let fetcher = AsyncProofFetcher::new(Arc::new(MockDbReaderWriter));
+        let mut expected_key_hashes = vec![];
+        for i in 0..10 {
+            let state_key = StateKey::Raw(format!("test_key_{}", i).into_bytes());
+            expected_key_hashes.push(state_key.hash());
+            let result = fetcher
+                .fetch_state_value_and_proof(&state_key, 0)
+                .expect("Should not fail.");
+            let expected_value = StateValue::from(match state_key {
+                StateKey::Raw(key) => key,
+                _ => unreachable!(),
+            });
+            assert_eq!(result.0, Some(expected_value));
+            assert!(result.1.is_none());
+        }
+
+        let proofs = fetcher.get_proof_cache();
+        assert_eq!(proofs.len(), 10);
```

Comment:
Doesn't the next assert imply this?

Reply:
Yes, but just want to be explicitly. It's not easy to read the error message if the next line prints out an array of 10 hash values is not equal to another array of 9 hash values, when they are unsorted.